git/06.linear-history.md

```diff
@@ -0,0 +1,71 @@
+## [Home](../README.md)
+
+## 6. Linear History
+
+At RoleModel we commend the practice of keeping git histories linear. This
+document discusses why.
+
+## Two approaches
+
+There are broadly two views on how to structure the history of a git repository:
+merging or rebasing. The merge view exhorts thorough reliance on git's
+unprecedented ability to create and use _merge commits_, which is simply a
+commit with more than one parent. The Linux kernel, which is perhaps the
+flagship git project because
+[it spawned git](https://git-scm.com/book/en/v2/Getting-Started-A-Short-History-of-Git) and
+because it has more activity than almost any other project, adheres to this
+view. Its maintainers have been known to create "octopus merges" with more than
+two parents; in fact,
+[one commit has 66 parents](https://www.destroyallsoftware.com/blog/2017/the-biggest-and-weirdest-commits-in-linux-kernel-git-history)!
+
+The other view is to rely on git's ability to _rebase_ commits onto another
+commit, which effectively replays the commits onto a new starting point. Using
+rebase allows us to have a _linear_ (or at least linearizable) history. You can
+see an example of what linear vs. non-linear history looks like in
+the [Merge Strategies](05.merge-strategies.md#background) best practice.
+
+As an aside, those who view the [purpose of git](01.purpose-of-git.md) as a log
+of what happened (rather than a logical history) eschew the use of history
+rewriting techniques like `git rebase`, and so take the merge view.
+
+## Recommendation
+
+At RoleModel we recommend linear history in almost all cases. The one exception
+is a very short-lived project that is unlikely to have anybody needing to
+understand the history.
+
+## Considerations
+
+The main tradeoff that we recognize is time spent creating the history vs. time
+spent reading and understanding the history. For long-lived projects, we
+optimize for the latter, just as we spend extra development time making the code
+legible for later readers. For short-lived projects, however, it may make sense
+to optimize for writing rather than reading.
+
+Another consideration is how the two approaches reconcile conflicts, which
+happen when git can't figure out how to bring two histories together. In the
+merge-based view, a conflict is surfaced during the merge. Resolving the
+conflict once is sufficient. In a rebase, however, there is sometimes a repeated
+conflict at every new commit that is being replayed. If you rebase before
+squashing, as we recommend, this can become annoying fast. Git's `rerere` option
+can help with this, which `re`plays `re`corded (conflict) `re`solutions.
```

Comment:
I was thinking about since we mentioned `rerere`.  Sometimes the issue in rebasing is that you are rebasing more than your commits.

Maybe it make sense to move `rerere` to the setup document.  And we link to Rebasing section of the Release branching doc